### PR TITLE
Finish the JSON clone algorithm for WebDriver

### PR DESCRIPTION
**Pull request by georgeroman**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23947)

### CODE REVIEW

**Comment by asajeffrey** [comment]
@jdm did you mean to r? yourself?

**Comment by jdm** [comment]
I did.

File: components/script/webdriver_handlers.rs [hunk 1]
```diff
@@ -199,6 +226,24 @@ pub unsafe fn jsval_to_webdriver(
...
+            if JS_CallFunctionName(
+                cx,
+                jsapi::Handle::from_marked_location(&*object),
+                name.as_ptr(),
+                &mut jsapi::HandleValueArray::new(),
+                jsapi::MutableHandle::from_marked_location(&mut value),
+            ) {
+                jsval_to_webdriver(cx, global_scope, Handle::new(&value))
+            } else {
+                Err(WebDriverJSError::UnknownType)
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: We should throw a DOM error here if JS_CallFunctionName failed.

File: components/script/webdriver_handlers.rs [hunk 2]
```diff
@@ -199,6 +226,24 @@ pub unsafe fn jsval_to_webdriver(
                     window.Document().upcast::<Node>().unique_id(),
                 )))
             }
+        } else if object_has_to_json_property(
+            cx,
+            global_scope,
+            jsapi::HandleObject::from_marked_location(&*object),
+        ) {
+            let name = CString::new("toJSON").unwrap();
+            let mut value = UndefinedValue();
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: Use rooted! here.

File: components/script/webdriver_handlers.rs [hunk 3]
```diff
@@ -199,6 +226,24 @@ pub unsafe fn jsval_to_webdriver(
...
+            if JS_CallFunctionName(
+                cx,
+                jsapi::Handle::from_marked_location(&*object),
+                name.as_ptr(),
+                &mut jsapi::HandleValueArray::new(),
+                jsapi::MutableHandle::from_marked_location(&mut value),
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: We can call value.handle_mut() instead.

File: components/script/webdriver_handlers.rs [hunk 4]
```diff
@@ -199,6 +226,24 @@ pub unsafe fn jsval_to_webdriver(
                     window.Document().upcast::<Node>().unique_id(),
                 )))
             }
+        } else if object_has_to_json_property(
+            cx,
+            global_scope,
+            jsapi::HandleObject::from_marked_location(&*object),
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: We can use object.handle() instead.

File: components/script/webdriver_handlers.rs [hunk 5]
```diff
@@ -199,6 +226,24 @@ pub unsafe fn jsval_to_webdriver(
...
+        ) {
+            let name = CString::new("toJSON").unwrap();
+            let mut value = UndefinedValue();
+            if JS_CallFunctionName(
+                cx,
+                jsapi::Handle::from_marked_location(&*object),
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: object.handle().

File: components/script/webdriver_handlers.rs [hunk 6]
```diff
@@ -245,7 +246,22 @@ pub unsafe fn jsval_to_webdriver(
                 Err(WebDriverJSError::UnknownType)
             }
         } else {
-            Err(WebDriverJSError::UnknownType)
+            let mut result = HashMap::new();
+
+            let common_properties = vec!["x", "y", "width", "height", "key"];
+            for property in common_properties.iter() {
+                rooted!(in(cx) let mut item = UndefinedValue());
+                if let Ok(_) = get_property_jsval(cx, object.handle(), property, item.handle_mut())
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: We should throw a DOM exception if this fails.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /webdriver/tests/execute_script/cyclic.py:
  └ PASS [expected FAIL] test_object

  ▶ Unexpected subtest result in /webdriver/tests/execute_script/cyclic.py:
  └ PASS [expected FAIL] test_array_in_object
```

**Pull request merged**